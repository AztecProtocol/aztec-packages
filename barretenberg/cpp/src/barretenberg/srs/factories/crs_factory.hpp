#pragma once
#include "barretenberg/common/mem.hpp"
#include "barretenberg/ecc/curves/bn254/bn254.hpp"
#include "barretenberg/ecc/curves/bn254/g1.hpp"
#include "barretenberg/ecc/curves/bn254/g2.hpp"
#include "barretenberg/ecc/curves/grumpkin/grumpkin.hpp"
#include <cstddef>

namespace bb::pairing {
struct miller_lines;
} // namespace bb::pairing

namespace bb::srs::factories {

/**
 * A prover crs representation.
 */
template <typename Curve> class ProverCrs {
  public:
    virtual ~ProverCrs() = default;
    ;

    /**
     *  @brief Returns the monomial points in a form to be consumed by scalar_multiplication pippenger algorithm.
     */
    virtual typename Curve::AffineElement* get_monomial_points() = 0;
    virtual size_t get_monomial_size() const = 0;
};

template <typename Curve> class VerifierCrs {
  public:
    virtual ~VerifierCrs() = default;
};
template <> class VerifierCrs<curve::BN254> {
    using Curve = curve::BN254;

  public:
    virtual Curve::G2AffineElement get_g2x() const = 0;
    /**
     * @brief As the G_2 element of the CRS is fixed, we can precompute the operations performed on it during the
     * pairing algorithm to optimize pairing computations.
     */
    virtual bb::pairing::miller_lines const* get_precomputed_g2_lines() const = 0;
    /**
     *  @brief Returns the first G_1 element from the CRS, used by the Shplonk verifier to compute the final
     * commtiment.
     */
    virtual Curve::AffineElement get_first_g1() const = 0;
};

template <> class VerifierCrs<curve::Grumpkin> {
    using Curve = curve::Grumpkin;

  public:
    /**
     * @brief Returns the G_1 elements in the CRS after the pippenger point table has been applied on them
     *
     */
    virtual Curve::AffineElement* get_monomial_points() const = 0;
    virtual size_t get_monomial_size() const = 0;
    /**
     * @brief Returns the first G_1 element from the CRS, used by the Shplonk verifier to compute the final
     * commtiment.
     */
    virtual Curve::AffineElement get_first_g1() const = 0;
};

/**
 * A factory class to return the prover crs and verifier crs on request.
 * You can construct an empty placeholder factory, because composers need to be given a factory at construction time.
 */
template <typename Curve> class CrsFactory {
  public:
    CrsFactory() = default;
    CrsFactory(CrsFactory&& other) = default;
    virtual ~CrsFactory() = default;
    virtual std::shared_ptr<bb::srs::factories::ProverCrs<Curve>> get_prover_crs(size_t) { return nullptr; }
    virtual std::shared_ptr<bb::srs::factories::VerifierCrs<Curve>> get_verifier_crs([[maybe_unused]] size_t degree = 0)
    {
        return nullptr;
    }
};

} // namespace bb::srs::factories