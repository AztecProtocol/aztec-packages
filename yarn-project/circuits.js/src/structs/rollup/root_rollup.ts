import { Fr } from '@aztec/foundation/fields';
import { BufferReader, Tuple } from '@aztec/foundation/serialize';

import {
  HISTORIC_BLOCKS_TREE_HEIGHT,
  L1_TO_L2_MSG_SUBTREE_SIBLING_PATH_LENGTH,
  NUMBER_OF_L1_L2_MESSAGES_PER_ROLLUP,
} from '../../cbind/constants.gen.js';
import { FieldsOf } from '../../utils/jsUtils.js';
import { serializeToBuffer } from '../../utils/serialize.js';
import { AggregationObject } from '../aggregation_object.js';
import { GlobalVariables } from '../global_variables.js';
import { AppendOnlyTreeSnapshot } from './append_only_tree_snapshot.js';
import { PreviousRollupData } from './previous_rollup_data.js';

/**
 * Represents inputs of the root rollup circuit.
 */
export class RootRollupInputs {
  constructor(
    /**
     * The previous rollup data.
     * Note: Root rollup circuit is the latest circuit the chain of circuits and the previous rollup data is the data
     * from 2 merge or base rollup circuits.
     */
    public previousRollupData: [PreviousRollupData, PreviousRollupData],
    /**
     * New L1 to L2 messages.
     */
    public newL1ToL2Messages: Tuple<Fr, typeof NUMBER_OF_L1_L2_MESSAGES_PER_ROLLUP>,
    /**
     * Sibling path of the new L1 to L2 message tree root.
     */
    public newL1ToL2MessagesTreeRootSiblingPath: Tuple<Fr, typeof L1_TO_L2_MSG_SUBTREE_SIBLING_PATH_LENGTH>,
    /**
     * Snapshot of the L1 to L2 message tree at the start of the rollup.
     */
    public startL1ToL2MessagesTreeSnapshot: AppendOnlyTreeSnapshot,
    /**
     * Snapshot of the historic block roots tree at the start of the rollup.
     */
    public startHistoricBlocksTreeSnapshot: AppendOnlyTreeSnapshot,
    /**
     * Sibling path of the new historic block roots tree root.
     */
    public newHistoricBlocksTreeSiblingPath: Tuple<Fr, typeof HISTORIC_BLOCKS_TREE_HEIGHT>,
  ) {}

  toBuffer() {
    return serializeToBuffer(...RootRollupInputs.getFields(this));
  }

  static from(fields: FieldsOf<RootRollupInputs>): RootRollupInputs {
    return new RootRollupInputs(...RootRollupInputs.getFields(fields));
  }

  static getFields(fields: FieldsOf<RootRollupInputs>) {
    return [
      fields.previousRollupData,
      fields.newL1ToL2Messages,
      fields.newL1ToL2MessagesTreeRootSiblingPath,
      fields.startL1ToL2MessagesTreeSnapshot,
      fields.startHistoricBlocksTreeSnapshot,
      fields.newHistoricBlocksTreeSiblingPath,
    ] as const;
  }
}

/**
 * Represents public inputs of the root rollup circuit.
 *
 * NOTE: in practice, we'll hash all of this up into a single public input, for cheap on-chain verification.
 */
export class RootRollupPublicInputs {
  constructor(
    /**
     * Native aggregation state at the end of the rollup.
     */
    public endAggregationObject: AggregationObject,

    /**
     * Global variables of the L2 block.
     */
    public globalVariables: GlobalVariables,
    /**
     * Snapshot of the private data tree at the start of the rollup.
     */
    public startPrivateDataTreeSnapshot: AppendOnlyTreeSnapshot,

    /**
     * Snapshot of the private data tree at the end of the rollup.
     */
    public endPrivateDataTreeSnapshot: AppendOnlyTreeSnapshot,

    /**
     * Snapshot of the nullifier tree at the start of the rollup.
     */
    public startNullifierTreeSnapshot: AppendOnlyTreeSnapshot,
    /**
     * Snapshot of the nullifier tree at the end of the rollup.
     */
    public endNullifierTreeSnapshot: AppendOnlyTreeSnapshot,

    /**
     * Snapshot of the contract tree at the start of the rollup.
     */
    public startContractTreeSnapshot: AppendOnlyTreeSnapshot,
    /**
     * Snapshot of the contract tree at the end of the rollup.
     */
    public endContractTreeSnapshot: AppendOnlyTreeSnapshot,

    /**
     * Root of the public data tree at the start of the rollup.
     */
    public startPublicDataTreeRoot: Fr,
    /**
     * Root of the public data tree at the end of the rollup.
     */
    public endPublicDataTreeRoot: Fr,

    /**
     * Snapshot of the historic private data tree roots tree at the start of the rollup.
     */
    public startTreeOfHistoricPrivateDataTreeRootsSnapshot: AppendOnlyTreeSnapshot,
    /**
     * Snapshot of the historic private data tree roots tree at the end of the rollup.
     */
    public endTreeOfHistoricPrivateDataTreeRootsSnapshot: AppendOnlyTreeSnapshot,

    /**
     * Snapshot of the historic contract tree roots tree at the start of the rollup.
     */
    public startTreeOfHistoricContractTreeRootsSnapshot: AppendOnlyTreeSnapshot,
    /**
     * Snapshot of the historic contract tree roots tree at the end of the rollup.
     */
    public endTreeOfHistoricContractTreeRootsSnapshot: AppendOnlyTreeSnapshot,

    /**
     * Snapshot of the L1 to L2 message tree at the start of the rollup.
     */
    public startL1ToL2MessagesTreeSnapshot: AppendOnlyTreeSnapshot,
    /**
     * Snapshot of the L1 to L2 message tree at the end of the rollup.
     */
    public endL1ToL2MessagesTreeSnapshot: AppendOnlyTreeSnapshot,

    /**
     * Snapshot of the historic L1 to L2 message tree roots tree at the start of the rollup.
     */
    public startTreeOfHistoricL1ToL2MessagesTreeRootsSnapshot: AppendOnlyTreeSnapshot,
    /**
     * Snapshot of the historic L1 to L2 message tree roots tree at the end of the rollup.
     */
    public endTreeOfHistoricL1ToL2MessagesTreeRootsSnapshot: AppendOnlyTreeSnapshot,

    /**
     * Snapshot of the historic blocks tree roots tree at the start of the rollup.
     */
    public startHistoricBlocksTreeSnapshot: AppendOnlyTreeSnapshot,
    /**
     * Snapshot of the historic blocks tree roots tree at the end of the rollup.
     */
    public endHistoricBlocksTreeSnapshot: AppendOnlyTreeSnapshot,

    /**
     * Hash of the calldata.
     */
    public calldataHash: [Fr, Fr],
    /**
     * Hash of the L1 to L2 messages.
     */
    public l1ToL2MessagesHash: [Fr, Fr],
  ) {}

  static getFields(fields: FieldsOf<RootRollupPublicInputs>) {
    return [
      fields.endAggregationObject,
      fields.globalVariables,
      fields.startPrivateDataTreeSnapshot,
      fields.endPrivateDataTreeSnapshot,
      fields.startNullifierTreeSnapshot,
      fields.endNullifierTreeSnapshot,
      fields.startContractTreeSnapshot,
      fields.endContractTreeSnapshot,
      fields.startPublicDataTreeRoot,
      fields.endPublicDataTreeRoot,
      fields.startTreeOfHistoricPrivateDataTreeRootsSnapshot,
      fields.endTreeOfHistoricPrivateDataTreeRootsSnapshot,
      fields.startTreeOfHistoricContractTreeRootsSnapshot,
      fields.endTreeOfHistoricContractTreeRootsSnapshot,
      fields.startL1ToL2MessagesTreeSnapshot,
      fields.endL1ToL2MessagesTreeSnapshot,
      fields.startTreeOfHistoricL1ToL2MessagesTreeRootsSnapshot,
      fields.endTreeOfHistoricL1ToL2MessagesTreeRootsSnapshot,
      fields.startHistoricBlocksTreeSnapshot,
      fields.endHistoricBlocksTreeSnapshot,
      fields.calldataHash,
      fields.l1ToL2MessagesHash,
    ] as const;
  }

  toBuffer() {
    return serializeToBuffer(...RootRollupPublicInputs.getFields(this));
  }

  static from(fields: FieldsOf<RootRollupPublicInputs>): RootRollupPublicInputs {
    return new RootRollupPublicInputs(...RootRollupPublicInputs.getFields(fields));
  }

  /**
   * Returns the sha256 hash of the calldata.
   * @returns The sha256 hash of the calldata.
   */
  public sha256CalldataHash(): Buffer {
    const high = this.calldataHash[0].toBuffer();
    const low = this.calldataHash[1].toBuffer();

    const hash = Buffer.alloc(32);
    for (let i = 0; i < 16; i++) {
      hash[i] = high[i + 16];
      hash[i + 16] = low[i + 16];
    }

    return hash;
  }

  /**
   * Deserializes a buffer into a `RootRollupPublicInputs` object.
   * @param buffer - The buffer to deserialize.
   * @returns The deserialized `RootRollupPublicInputs` object.
   */
  public static fromBuffer(buffer: Buffer | BufferReader): RootRollupPublicInputs {
    const reader = BufferReader.asReader(buffer);
    return new RootRollupPublicInputs(
      reader.readObject(AggregationObject),
      reader.readObject(GlobalVariables),
      reader.readObject(AppendOnlyTreeSnapshot),
      reader.readObject(AppendOnlyTreeSnapshot),
      reader.readObject(AppendOnlyTreeSnapshot),
      reader.readObject(AppendOnlyTreeSnapshot),
      reader.readObject(AppendOnlyTreeSnapshot),
      reader.readObject(AppendOnlyTreeSnapshot),
      reader.readFr(),
      reader.readFr(),
      reader.readObject(AppendOnlyTreeSnapshot),
      reader.readObject(AppendOnlyTreeSnapshot),
      reader.readObject(AppendOnlyTreeSnapshot),
      reader.readObject(AppendOnlyTreeSnapshot),
      reader.readObject(AppendOnlyTreeSnapshot),
      reader.readObject(AppendOnlyTreeSnapshot),
      reader.readObject(AppendOnlyTreeSnapshot),
      reader.readObject(AppendOnlyTreeSnapshot),
      reader.readObject(AppendOnlyTreeSnapshot),
      reader.readObject(AppendOnlyTreeSnapshot),
      [reader.readFr(), reader.readFr()],
      [reader.readFr(), reader.readFr()],
    );
  }
}
