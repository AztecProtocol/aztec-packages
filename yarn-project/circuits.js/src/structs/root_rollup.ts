import { assertLength, FieldsOf } from '../utils/jsUtils.js';
import { serializeToBuffer } from '../utils/serialize.js';
import { AppendOnlyTreeSnapshot } from './base_rollup.js';
import { CONTRACT_TREE_HEIGHT, NULLIFIER_TREE_HEIGHT, PRIVATE_DATA_TREE_HEIGHT } from './constants.js';
import { PreviousRollupData } from './merge_rollup.js';
import { AggregationObject, Fr } from './shared.js';

export class RootRollupInputs {
  constructor(
    public previousRollupData: [PreviousRollupData, PreviousRollupData],

    public startPrivateDataTreeSnapshot: AppendOnlyTreeSnapshot,
    // Note: the start_nullifier_tree_snapshot is contained within previous_rollup_data[0].public_inputs.start_nullifier_tree_snapshot.
    public startContractTreeSnapshot: AppendOnlyTreeSnapshot,

    // For inserting the new subtrees into their respective trees:
    // Note: the insertion leaf index can be derived from the above snapshots' `next_available_leaf_index` values.
    public newCommitmentsSubtreeSiblingPath: Fr[],
    public newNullifiersSubtreeSiblingPath: Fr[],
    public newContractsSubtreeSiblingPath: Fr[],

    public proverId: Fr,
  ) {
    assertLength(this, 'newCommitmentsSubtreeSiblingPath', PRIVATE_DATA_TREE_HEIGHT);
    assertLength(this, 'newNullifiersSubtreeSiblingPath', NULLIFIER_TREE_HEIGHT);
    assertLength(this, 'newContractsSubtreeSiblingPath', CONTRACT_TREE_HEIGHT);
  }

  toBuffer() {
    return serializeToBuffer(
      this.previousRollupData,
      this.startPrivateDataTreeSnapshot,
      this.startContractTreeSnapshot,
      this.newCommitmentsSubtreeSiblingPath,
      this.newNullifiersSubtreeSiblingPath,
      this.newContractsSubtreeSiblingPath,
      this.proverId,
    );
  }
}

export class RootRollupPublicInputs {
  constructor(
    // NOTE: in practice, we'll hash all of this up into a single public input, for cheap on-chain verification.
    public endAggregationObject: AggregationObject,

    // constants: ConstantRollupData // TODO maybe don't include this

    public startPrivateDataTreeSnapshot: AppendOnlyTreeSnapshot,
    public endPrivateDataTreeSnapshot: AppendOnlyTreeSnapshot,

    public startNullifierTreeSnapshot: AppendOnlyTreeSnapshot,
    public endNullifierTreeSnapshot: AppendOnlyTreeSnapshot,

    public startContractTreeSnapshot: AppendOnlyTreeSnapshot,
    public endContractTreeSnapshot: AppendOnlyTreeSnapshot,

    public startTreeOfHistoricPrivateDataTreeRootsSnapshot: AppendOnlyTreeSnapshot,
    public endTreeOfHistoricPrivateDataTreeRootsSnapshot: AppendOnlyTreeSnapshot,

    public startTreeOfHistoricContractTreeRootsSnapshot: AppendOnlyTreeSnapshot,
    public endTreeOfHistoricContractTreeRootsSnapshot: AppendOnlyTreeSnapshot,

    // Hashes (probably sha256) to make public inputs constant-sized
    // (to then be unpacked on-chain)
    // UPDATE: we should instead just hash all of the below into a single value. See big diagram of sha256 hashing bottom-right of here.
    // TODO: I've put `fr`, but these hash values' types might need to be two fields if we want all 256-bits, for security purposes.
    public newCommitmentsHash: Fr,
    public newNullifiersHash: Fr,
    public newL1MsgsHash: Fr,
    public newContractDataHash: Fr,
    public proverContributionsHash: Fr, // TODO: spec how funds are distributed to provers.
  ) {}

  static getFields(fields: FieldsOf<RootRollupPublicInputs>) {
    return [
      fields.endAggregationObject,
      fields.startPrivateDataTreeSnapshot,
      fields.endPrivateDataTreeSnapshot,
      fields.startNullifierTreeSnapshot,
      fields.endNullifierTreeSnapshot,
      fields.startContractTreeSnapshot,
      fields.endContractTreeSnapshot,
      fields.startTreeOfHistoricPrivateDataTreeRootsSnapshot,
      fields.endTreeOfHistoricPrivateDataTreeRootsSnapshot,
      fields.startTreeOfHistoricContractTreeRootsSnapshot,
      fields.endTreeOfHistoricContractTreeRootsSnapshot,
      fields.newCommitmentsHash,
      fields.newNullifiersHash,
      fields.newL1MsgsHash,
      fields.newContractDataHash,
      fields.proverContributionsHash,
    ] as const;
  }

  toBuffer() {
    return serializeToBuffer(...RootRollupPublicInputs.getFields(this));
  }

  static from(fields: FieldsOf<RootRollupPublicInputs>): RootRollupPublicInputs {
    return new RootRollupPublicInputs(...RootRollupPublicInputs.getFields(fields));
  }
}
