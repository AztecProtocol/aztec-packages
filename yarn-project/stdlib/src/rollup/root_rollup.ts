import { BlockBlobPublicInputs } from '@aztec/blob-lib';
import { AZTEC_MAX_EPOCH_DURATION } from '@aztec/constants';
import { makeTuple } from '@aztec/foundation/array';
import { Fr } from '@aztec/foundation/fields';
import { bufferSchemaFor } from '@aztec/foundation/schemas';
import { BufferReader, type Tuple, serializeToBuffer, serializeToFields } from '@aztec/foundation/serialize';
import { bufferToHex, hexToBuffer } from '@aztec/foundation/string';
import type { FieldsOf } from '@aztec/foundation/types';

import { FeeRecipient } from './block_root_or_block_merge_public_inputs.js';
import { PreviousRollupBlockData } from './previous_rollup_block_data.js';

/**
 * Represents inputs of the root rollup circuit.
 */
export class RootRollupInputs {
  constructor(
    /**
     * The previous rollup data.
     * Note: Root rollup circuit is the latest circuit the chain of circuits and the previous rollup data is the data
     * from 2 block merge circuits.
     */
    public previousRollupData: [PreviousRollupBlockData, PreviousRollupBlockData],
    /** Identifier of the prover for this root rollup. */
    public proverId: Fr,
  ) {}

  /**
   * Serializes the inputs to a buffer.
   * @returns - The inputs serialized to a buffer.
   */
  toBuffer() {
    return serializeToBuffer(...RootRollupInputs.getFields(this));
  }

  /**
   * Serializes the inputs to a hex string.
   * @returns The instance serialized to a hex string.
   */
  toString() {
    return bufferToHex(this.toBuffer());
  }

  /**
   * Creates a new instance from fields.
   * @param fields - Fields to create the instance from.
   * @returns A new RootRollupInputs instance.
   */
  static from(fields: FieldsOf<RootRollupInputs>): RootRollupInputs {
    return new RootRollupInputs(...RootRollupInputs.getFields(fields));
  }

  /**
   * Extracts fields from an instance.
   * @param fields - Fields to create the instance from.
   * @returns An array of fields.
   */
  static getFields(fields: FieldsOf<RootRollupInputs>) {
    return [fields.previousRollupData, fields.proverId] as const;
  }

  /**
   * Deserializes the inputs from a buffer.
   * @param buffer - A buffer to deserialize from.
   * @returns A new RootRollupInputs instance.
   */
  static fromBuffer(buffer: Buffer | BufferReader): RootRollupInputs {
    const reader = BufferReader.asReader(buffer);
    return new RootRollupInputs(
      [reader.readObject(PreviousRollupBlockData), reader.readObject(PreviousRollupBlockData)],
      Fr.fromBuffer(reader),
    );
  }

  /**
   * Deserializes the inputs from a hex string.
   * @param str - A hex string to deserialize from.
   * @returns A new RootRollupInputs instance.
   */
  static fromString(str: string) {
    return RootRollupInputs.fromBuffer(hexToBuffer(str));
  }

  /** Returns a representation for JSON serialization. */
  toJSON() {
    return this.toBuffer();
  }

  /** Creates an instance from a string. */
  static get schema() {
    return bufferSchemaFor(RootRollupInputs);
  }
}

/**
 * Represents public inputs of the root rollup circuit.
 *
 * NOTE: in practice, we'll hash all of this up into a single public input, for cheap on-chain verification.
 */
export class RootRollupPublicInputs {
  constructor(
    /** Root of the archive tree before this rollup is processed */
    public previousArchiveRoot: Fr,
    /** Root of the archive tree after this rollup is processed */
    public endArchiveRoot: Fr,
    public proposedBlockHeaderHashes: Tuple<Fr, typeof AZTEC_MAX_EPOCH_DURATION>,
    public fees: Tuple<FeeRecipient, typeof AZTEC_MAX_EPOCH_DURATION>,
    public chainId: Fr,
    public version: Fr,
    public vkTreeRoot: Fr,
    public protocolContractTreeRoot: Fr,
    public proverId: Fr,
    public blobPublicInputs: Tuple<BlockBlobPublicInputs, typeof AZTEC_MAX_EPOCH_DURATION>,
  ) {}

  static getFields(fields: FieldsOf<RootRollupPublicInputs>) {
    return [
      fields.previousArchiveRoot,
      fields.endArchiveRoot,
      fields.proposedBlockHeaderHashes,
      fields.fees,
      fields.chainId,
      fields.version,
      fields.vkTreeRoot,
      fields.protocolContractTreeRoot,
      fields.proverId,
      fields.blobPublicInputs,
    ] as const;
  }

  toBuffer() {
    return serializeToBuffer(...RootRollupPublicInputs.getFields(this));
  }

  toFields(): Fr[] {
    return serializeToFields(...RootRollupPublicInputs.getFields(this));
  }

  static from(fields: FieldsOf<RootRollupPublicInputs>): RootRollupPublicInputs {
    return new RootRollupPublicInputs(...RootRollupPublicInputs.getFields(fields));
  }

  /**
   * Deserializes a buffer into a `RootRollupPublicInputs` object.
   * @param buffer - The buffer to deserialize.
   * @returns The deserialized `RootRollupPublicInputs` object.
   */
  public static fromBuffer(buffer: Buffer | BufferReader): RootRollupPublicInputs {
    const reader = BufferReader.asReader(buffer);
    return new RootRollupPublicInputs(
      Fr.fromBuffer(reader),
      Fr.fromBuffer(reader),
      reader.readArray(AZTEC_MAX_EPOCH_DURATION, Fr),
      reader.readArray(AZTEC_MAX_EPOCH_DURATION, FeeRecipient),
      Fr.fromBuffer(reader),
      Fr.fromBuffer(reader),
      Fr.fromBuffer(reader),
      Fr.fromBuffer(reader),
      Fr.fromBuffer(reader),
      reader.readArray(AZTEC_MAX_EPOCH_DURATION, BlockBlobPublicInputs),
    );
  }

  toString() {
    return bufferToHex(this.toBuffer());
  }

  static fromString(str: string) {
    return RootRollupPublicInputs.fromBuffer(hexToBuffer(str));
  }

  /** Returns a representation for JSON serialization. */
  toJSON() {
    return this.toBuffer();
  }

  /** Creates an instance from a string. */
  static get schema() {
    return bufferSchemaFor(RootRollupPublicInputs);
  }

  /** Creates a random instance. */
  static random() {
    return new RootRollupPublicInputs(
      Fr.random(),
      Fr.random(),
      makeTuple(AZTEC_MAX_EPOCH_DURATION, Fr.random),
      makeTuple(AZTEC_MAX_EPOCH_DURATION, FeeRecipient.random),
      Fr.random(),
      Fr.random(),
      Fr.random(),
      Fr.random(),
      Fr.random(),
      makeTuple(AZTEC_MAX_EPOCH_DURATION, BlockBlobPublicInputs.empty),
    );
  }
}
