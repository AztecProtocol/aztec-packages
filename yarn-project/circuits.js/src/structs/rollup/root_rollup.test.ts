import { expectReserializeToMatchObject, expectSerializeToMatchSnapshot } from '../../tests/expectSerialize.js';
import { makeRootRollupInputs, makeRootRollupPublicInputs } from '../../tests/factories.js';
import { RootRollupPublicInputs } from './root_rollup.js';

describe('structs/root_rollup', () => {
  it(`serializes a RootRollupInput and prints it`, async () => {
    await expectSerializeToMatchSnapshot(
      makeRootRollupInputs().toBuffer(),
      'abis__test_roundtrip_serialize_root_rollup_inputs',
    );
  });

  it(`serializes a RootRollupPublicInputs and prints it`, async () => {
    await expectSerializeToMatchSnapshot(
      makeRootRollupPublicInputs().toBuffer(),
      'abis__test_roundtrip_serialize_root_rollup_public_inputs',
    );
  });

  it(`serializes a RootRollupPublicInputs and deserializes it back`, async () => {
    await expectReserializeToMatchObject(
      makeRootRollupPublicInputs(),
      'abis__test_roundtrip_reserialize_root_rollup_public_inputs',
      RootRollupPublicInputs.fromBuffer,
    );
  });
});
