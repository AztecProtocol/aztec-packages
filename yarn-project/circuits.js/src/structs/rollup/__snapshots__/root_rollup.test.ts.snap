// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`structs/root_rollup serializes a RootRollupInput and prints it 1`] = `
"previous_rollup_data: [ base_or_merge_rollup_public_inputs:
rollup_type: 0
rollup_subtree_height: 0x0
end_aggregation_object: P0: { 0x100, 0x101 }
P1: { 0x200, 0x201 }
public_inputs: [
 0x102
 0x103
 0x104
 0x105
]
proof_witness_indices: [ 106 107 108 109 10a 10b ]
has_data: 0

constants: start_historic_blocks_tree_roots_snapshot:
root: 0x500
next_available_leaf_index: 500

private_kernel_vk_tree_root: 0x601
public_kernel_vk_tree_root: 0x602
base_rollup_vk_hash: 0x603
merge_rollup_vk_hash: 0x604
global_variables: chain_id: 0x605
version: 0x606
block_number: 0x607
timestamp: 0x608


start_private_data_tree_snapshot: root: 0x300
next_available_leaf_index: 300

end_private_data_tree_snapshot: root: 0x400
next_available_leaf_index: 400

start_nullifier_tree_snapshot: root: 0x500
next_available_leaf_index: 500

end_nullifier_tree_snapshot: root: 0x600
next_available_leaf_index: 600

start_contract_tree_snapshot: root: 0x700
next_available_leaf_index: 700

end_contract_tree_snapshot: root: 0x800
next_available_leaf_index: 800

start_public_data_tree_root: 0x900
end_public_data_tree_root: 0x1000
calldata_hash: [ 0x901 0x902 ]

proof: [ 50 50 50 50 50 50 50 50 50 50 50 50 50 50 50 50 ]
vk: key.circuit_type: 0
key.circuit_size: 65
key.num_public_inputs: 66
key.commitments: [
 A: { 0x200, 0x300 }
]
key.contains_recursive_proof: 1
key.recursive_proof_public_input_indices: [ 190 191 192 193 194 ]

vk_index: 110
vk_sibling_path: leaf_index: 0x120
sibling_path: [ 0x120 0x121 0x122 0x123 0x124 0x125 0x126 0x127 ]

 base_or_merge_rollup_public_inputs:
rollup_type: 0
rollup_subtree_height: 0x0
end_aggregation_object: P0: { 0x1100, 0x1101 }
P1: { 0x1200, 0x1201 }
public_inputs: [
 0x1102
 0x1103
 0x1104
 0x1105
]
proof_witness_indices: [ 1106 1107 1108 1109 110a 110b ]
has_data: 0

constants: start_historic_blocks_tree_roots_snapshot:
root: 0x1500
next_available_leaf_index: 1500

private_kernel_vk_tree_root: 0x1601
public_kernel_vk_tree_root: 0x1602
base_rollup_vk_hash: 0x1603
merge_rollup_vk_hash: 0x1604
global_variables: chain_id: 0x1605
version: 0x1606
block_number: 0x1607
timestamp: 0x1608


start_private_data_tree_snapshot: root: 0x1300
next_available_leaf_index: 1300

end_private_data_tree_snapshot: root: 0x1400
next_available_leaf_index: 1400

start_nullifier_tree_snapshot: root: 0x1500
next_available_leaf_index: 1500

end_nullifier_tree_snapshot: root: 0x1600
next_available_leaf_index: 1600

start_contract_tree_snapshot: root: 0x1700
next_available_leaf_index: 1700

end_contract_tree_snapshot: root: 0x1800
next_available_leaf_index: 1800

start_public_data_tree_root: 0x1900
end_public_data_tree_root: 0x2000
calldata_hash: [ 0x1901 0x1902 ]

proof: [ 50 50 50 50 50 50 50 50 50 50 50 50 50 50 50 50 ]
vk: key.circuit_type: 0
key.circuit_size: 65
key.num_public_inputs: 66
key.commitments: [
 A: { 0x200, 0x300 }
]
key.contains_recursive_proof: 1
key.recursive_proof_public_input_indices: [ 190 191 192 193 194 ]

vk_index: 1110
vk_sibling_path: leaf_index: 0x1120
sibling_path: [ 0x1120 0x1121 0x1122 0x1123 0x1124 0x1125 0x1126 0x1127 ]

 ]
new_l1_to_l2_messages: [ 0x2100 0x2101 0x2102 0x2103 0x2104 0x2105 0x2106 0x2107 0x2108 0x2109 0x210a 0x210b 0x210c 0x210d 0x210e 0x210f ]
new_l1_to_l2_messages_tree_root_sibling_path: [ 0x2100 0x2101 0x2102 0x2103 0x2104 0x2105 0x2106 0x2107 0x2108 0x2109 0x210a 0x210b ]
start_l1_to_l2_messages_tree_snapshot: root: 0x2200
next_available_leaf_index: 8704

start_historic_blocks_tree_snapshot: root: 0x2200
next_available_leaf_index: 8704

new_historic_blocks_tree_sibling_path: [ 0x2400 0x2401 0x2402 0x2403 0x2404 0x2405 0x2406 0x2407 0x2408 0x2409 0x240a 0x240b 0x240c 0x240d 0x240e 0x240f ]
"
`;

exports[`structs/root_rollup serializes a RootRollupPublicInputs and prints it 1`] = `
"end_aggregation_object:
P0: { 0x0, 0x1 }
P1: { 0x100, 0x101 }
public_inputs: [
 0x2
 0x3
 0x4
 0x5
]
proof_witness_indices: [ 6 7 8 9 10 11 ]
has_data: 0

global_variables:
chain_id: 0x100
version: 0x101
block_number: 0x102
timestamp: 0x103

start_private_data_tree_snapshot:
root: 0x200
next_available_leaf_index: 512

end_private_data_tree_snapshot:
root: 0x300
next_available_leaf_index: 768

start_nullifier_tree_snapshot:
root: 0x400
next_available_leaf_index: 1024

end_nullifier_tree_snapshot:
root: 0x500
next_available_leaf_index: 1280

start_contract_tree_snapshot:
root: 0x600
next_available_leaf_index: 1536

end_contract_tree_snapshot:
root: 0x700
next_available_leaf_index: 1792

start_public_data_tree_root: 0x800
end_public_data_tree_root: 0x900
start_tree_of_historic_private_data_tree_roots_snapshot: root: 0xa00
next_available_leaf_index: 2560

end_tree_of_historic_private_data_tree_roots_snapshot: root: 0xb00
next_available_leaf_index: 2816

start_tree_of_historic_contract_tree_roots_snapshot: root: 0xc00
next_available_leaf_index: 3072

end_tree_of_historic_contract_tree_roots_snapshot: root: 0xd00
next_available_leaf_index: 3328

start_l1_to_l2_messages_tree_snapshot: root: 0xe00
next_available_leaf_index: 3584

end_l1_to_l2_messages_tree_snapshot: root: 0xf00
next_available_leaf_index: 3840

start_tree_of_historic_l1_to_l2_messages_tree_roots_snapshot: root: 0x1000
next_available_leaf_index: 4096

end_tree_of_historic_l1_to_l2_messages_tree_roots_snapshot: root: 0x1100
next_available_leaf_index: 4352

start_historic_blocks_tree_snapshot: root: 0x1200
next_available_leaf_index: 4608

end_historic_blocks_tree_snapshot: root: 0x1300
next_available_leaf_index: 4864

calldata_hash: [ 0x1 0x2 ]
l1_to_l2_messages_hash: [ 0x3 0x4 ]
"
`;
